sdks/java/io/xml/src/test/java/org/apache/beam/sdk/io/xml/XmlIOIT.java

```diff
@@ -0,0 +1,183 @@
+    100_000, "af7775de90d0b0c8bbc36273fbca26fe",
+    100_000_000, "bfee52b33aa1552b9c1bfa8bcc41ae80"
+  );
+
+  private static Integer numberOfRecords;
+
+  private static String filenamePrefix;
+
+  private static Charset charset;
+
+  @Rule
+  public TestPipeline pipeline = TestPipeline.create();
+
+  @Rule
+  public TestPipeline readPipeline = TestPipeline.create();
+
+  @BeforeClass
+  public static void setup() {
+    PipelineOptionsFactory.register(IOTestPipelineOptions.class);
+    IOTestPipelineOptions options = TestPipeline
+      .testingPipelineOptions()
+      .as(IOTestPipelineOptions.class);
+
+    filenamePrefix = appendTimestampSuffix(options.getFilenamePrefix());
+    numberOfRecords = options.getNumberOfRecords();
+    charset = Charset.forName(options.getCharset());
+  }
+
+  @Test
+  public void writeThenReadViaSinkAndReadFiles() {
+    PCollection<String> testFileNames = pipeline
+      .apply("Generate sequence", GenerateSequence.from(0).to(numberOfRecords))
+      .apply("Create Birds", MapElements.via(new LongToBird()))
+      .apply("Write birds to xml files", FileIO.<Bird>write()
+          .via(XmlIO.sink(Bird.class)
+            .withRootElement("birds")
+            .withCharset(charset))
+          .to(filenamePrefix)
+          .withPrefix("birds")
+          .withSuffix(".xml"))
+      .getPerDestinationOutputFilenames()
+      .apply("Prevent fusion", Reshuffle.viaRandomKey())
+      .apply("Get file names", Values.create());
+
+    PCollection<Bird> birds = testFileNames
+      .apply("Find files", FileIO.matchAll())
+      .apply("Read matched files", FileIO.readMatches())
+      .apply("Read xml files", XmlIO.<Bird>readFiles()
+        .withRecordClass(Bird.class).withRootElement("birds")
+        .withRecordElement("bird")
+        .withCharset(charset));
+
+    PCollection<String> consolidatedHashcode = birds
+      .apply("Map birds to strings", MapElements.via(new BirdToString()))
+      .apply("Calculate hashcode", Combine.globally(new HashingFn()));
+
+    String expectedHash = IOTestHelper.getHashForRecordCount(numberOfRecords, EXPECTED_HASHES);
+    PAssert.thatSingleton(consolidatedHashcode).isEqualTo(expectedHash);
+
+    testFileNames.apply("Delete test files", ParDo.of(new DeleteFileFn())
+        .withSideInputs(consolidatedHashcode.apply(View.asSingleton())));
+
+    pipeline.run().waitUntilFinish();
+  }
+
+  @Test
+  public void writeThenReadViaWriteAndRead() {
+    String filenamePattern = filenamePrefix + "*";
+
+    pipeline
+      .apply("Generate sequence", GenerateSequence.from(0).to(numberOfRecords))
+      .apply("Create Birds", MapElements.via(new LongToBird()))
+      .apply("Write birds to xml files", XmlIO.<Bird>write()
+          .to(filenamePrefix)
+          .withRecordClass(Bird.class)
+          .withRootElement("birds")
+          .withCharset(charset));
+
+    pipeline.run().waitUntilFinish();
+
+    PCollection<String> consolidatedHashcode = readPipeline
```

Comment:
<!--thread_id:cc_171385121_t; commit:e3960b9fea070f4d514b71629fad156fc4f8773c; resolved:0-->
<!--section:context-quote-->
> **lgajowy** wrote:
> I tested the two ways of using the XmlIO: 
>  - using FileIO with XmlIO.sink and then using XmlIO.readFiles() method (this involves FileIO usage but allows us to perform all operations on one pipeline)
>  - using write() and read() methods from XmlIO only (this needs two pipelines)
> 
> This (in my opinion) has the following benefits:
>  - tests both scenarios (which can be helpful in finding regressions)
>  - shows two ways of implementing read and write scenario
>  - can show the difference between execution times of the tests 
> 
> It was a low effort to me so I decided to leave both. :) If this is not needed I can delete the "two pipeline" one. Should I?

<!--section:body-->
I think we should simplify and go with the single pipeline option. There is not enough of a diff between XmlIO.<Bird>readFiles() and XmlIO.<Bird>read() to justify an additional test.